### ISSUE TITLE
re-design the architecture of FreshRSS

### ISSUE THREAD

**Issue by oupala** [FreshRSS/FreshRSS#956]
FreshRSS works well for desktop web client, for mobile web client, for mobile native client (such as EasyRSS).

This leads me to another question : do you have the plan to split FreshRSS in (at least) two modules : FreshRSS-API for the core of FreshRSS, and FreshRSS-GUI for the web interface.

We could then choose to install only the core (and use EasyRSS as the client), or both of of them, or the core on one server and the web interface on another server. We could even imagine install the web interface locally, on my browser (as it is possible now with html5 localStorage, which is hardly used by Firefox OS, but could also work on any recent browser).

It could be a good milestone for FreshRSS, as it allows to get smaller modules, easier to maintain, to understand, and that could evolve independently with more agility.

I don't know if the web interface is in the current version already using the FreshRSS API or not. If not, this would mean some refactoring on the web interface side so that the web interface do not use internal mechanism any more, and only rely on the public API of FreshRSS-API.

**Comment by marienfressinaud** [comment]
Well, it is a very good idea and I was considering this option but it means a lot of work. It should be considered for a v2 only but we don't have enough time for the moment to work on a such big milestone.

But thanks anyway for the idea! :)

**Comment by aledeg** [comment]
I'd like to add that we could also change the backbone of the application by using more libraries. I know that in the past you didn't want to rely on external libraries but if we use libraries well tested and used by millions, we could focus on the application and not on the libraries.
For instance, @Alkarex  worked hard to make FRSS postgres compliant where it might have been easier with some ORM like doctrine.

See #995

**Comment by Alkarex** [comment]
@aledeg I am of course all in favour of having better code :-) Indeed, a rewrite / re-organisation of several parts would be very nice.

For the architecture, one characteristic I would like to see is to have less data / copies in memory, and to use streams instead. It was the main reason exporting large datasets failed, and I did a bit of it in https://github.com/FreshRSS/FreshRSS/pull/1373

Regarding the libraries, I suggest discussing individual cases. I am personally not against good third-party libraries, but not if it is only for the sake of having libraries. I think we should pick them carefully, and mainly if they actually improve the product (e.g. faster, lighter, better features) or reduce the total code base. Third-party libraries can be nice but can also be a burden when they do not perform exactly as wanted, when they evolve in an unwanted direction, or when they are not supported anymore (e.g. SimplePie during a long period). There is also a learning cost, compared to vanilla PHP, HTML, JS, SQL.

For instance, doctrine does not support all DB features, and a simple SQL change can become a complex hack (random example https://stackoverflow.com/questions/41394814/doctrine-how-to-use-any-value-mysql-function ). My main concern would clearly be the ability to keep or improve the current SQL performance. Several queries were manually optimized, and I think SQL performance is an essential aspect of FreshRSS.
Adding PostgreSQL was not that bad, and would have been even easier if we had thought about using more standard SQL from day one. For reference, the following files are more or less the additional code for PostgreSQL (and I do not think all of it would be supported by Doctrine):
* https://github.com/FreshRSS/FreshRSS/blob/dev/app/Models/DatabaseDAOPGSQL.php
* https://github.com/FreshRSS/FreshRSS/blob/dev/app/Models/EntryDAOPGSQL.php
* https://github.com/FreshRSS/FreshRSS/blob/dev/app/Models/StatsDAOPGSQL.php
To be more concrete, I think an interesting test could be to implement the two-three main SQL requests (the main insert, and the main select) in Doctrine, see whether it can be written efficiently at all, and compare performances in MySQL, SQLite, PostgreSQL on a large DB. E.g.:
* https://github.com/FreshRSS/FreshRSS/blob/dev/app/Models/EntryDAOPGSQL.php#L26
* https://github.com/FreshRSS/FreshRSS/blob/dev/app/Models/EntryDAO.php#L749

Recently, I removed the favicon library and wrote a new code (shorter, fewer requests, more cases working) for similar reasons https://github.com/FreshRSS/FreshRSS/pull/1504

By the way, I am also contemplating providing an alternative to SimplePie, again for similar reasons (too heavy, too large, too much legacy, undocumented breaking changes). If I ever do so, my plan would be to have a transition period during which one can pick one or the other, and decide after a few versions which to keep.

**Comment by marienfressinaud** [comment]
Frédéric Guillot wrote https://github.com/fguillot/picoFeed for Miniflux which is used by Owncloud News too. It seems to be a great alternative :).

About using additional libraries, I agree with arguments of @Alkarex :)

**Comment by aledeg** [comment]
To be clear, I don't want to add libraries for the sake of adding libraries. But for instance, we are facing at the moment a problem with the logs. It might be solved with a log library that follows the PSR. And even if it doesn't solve the problem, we open our project to other dimensions.
I agree that we need to evaluate what solution is better for our use case.

**Comment by oupala** [comment]
My point in this issue was about about using libraries or not, but about how to redesign FreshRSS to add more flexibility to its architecture.

I recently view a good conference about Knative named [Serverless containers with Knative and Cloud Run](https://www.youtube.com/watch?v=PLFLTMHB5Do) by Mete Atamel. This conference is mainly about C# and Google Cloud Platform, and hence far beyond the point I want to talk about in this issue.

However, the [conference ends with the rewrite of a monolith that processes RSS feed](https://www.youtube.com/watch?v=PLFLTMHB5Do&t=41m28s), just as FreshRSS. Then name of the application is [amathus](https://github.com/meteatamel/amathus/). They explained their journey switching from a monolith to a micro-services architecture.

I find their [architecture design](https://github.com/meteatamel/amathus/blob/master/Amathus/Shared/architecture.png) very interesting:

![architecture design](https://raw.githubusercontent.com/meteatamel/amathus/master/Amathus/Shared/architecture.png)

We can see that their application has been splitted into several services:

- the **reader service** that is triggered periodically by a cron, and which is in charge of fetching the rss feeds
- the **converter service** that is listening to changes from the output of the reader service (they use a datastore to share content), and which is in charge of processing the fetched feeds and the articles (such as filtering, tagging, or whatever)
- the **web service** that is listening to changes from the output of the converter service (they use a datastore to share content), and which is in charge of serving web pages to the user

In the case of FreshRSS, the app can be used as a web application, or as a REST api for client side application on desktop or on smartphone. It would required to add a fourth service to end with a total of 4 services :

- the **reader service** that is triggered periodically by a cron, and which is in charge of fetching the rss feeds
- the **converter service** that is listening to changes from the output of the reader service (they use a datastore to share content), and which is in charge of processing the fetched feeds and the articles (such as filtering, tagging, or whatever)
- the **api service** that is always on, and is listening to changes from the output of the converter service (they use a datastore to share content), and which is in charge of responding to api calls, either from the web service, of from a desktop or a smartphone client such as EasyRSS (see [complete list](https://github.com/FreshRSS/FreshRSS#apis--native-apps))
- the **web service** that is always on, and is using the api provided by the api service, and which is in charge of serving web pages to the user for those who are using the web interface of FreshRSS

This would ease the scaling of FreshRSS where we could scale some services only, and not the whole application.

We're not even obliged to run each service independantly as it could be possible to run all service on the same instance, I'm more thinking at an architecture that could easily switch from a [modular monolith](https://www.youtube.com/watch?v=F8C_iPGhHoI) that could be deployed as microservices, or as container, or as functions as a service...

I'm harldy waiting for your feedback about my thoughts.

**Comment by marienfressinaud** [comment]
TL;DR; if we had to split FRSS in several services, the more interesting one would be the "reader service", but it should not be done at the expense of easy installation for amateur adminsys. Also monolith is beautiful :heart: 

Thanks for your suggestion @oupala which looks logic to me and could be a good basis for an improvement of the architecture.

However, while I'm in favor of a better architecture, I would warn against going to micro-services (or even modular monolith) and I rarely see in these conferences drawbacks of this kind of architecture. In fact, I don't understand why monolith applications have such a bad reputation: it's simpler to develop, simpler to understand, simpler to debug, simpler to install, etc. I think we both agree with @Alkarex that we want FreshRSS to be easily installable on simple infrastructures and that it should not be necessary to have strong adminsys skills to manage it. Just to be clear: I also understand the attractiveness of micro-services, but it should be driven by specific needs. I think micro-services are a solution to a problem that FreshRSS doesn't really have (i.e. scaling).

That being said, it's not completely true. I'm hosting one of the biggest instance of FreshRSS with > 200 users. Fortunately, not all of them are active and so, I only fetch articles for a handful of users (i.e. the "reader service"). However, after an upgrade the other day, I accidentally re-enabled all the users and I faced performance issues in fetching, which impacted the webservice. The solution was to disable fetching for these users, but let's consider they're active users (which could be easily the case), next solutions are:

1. to upgrade the server with more memory and CPU
2. to extract a "reader service" on a dedicated server so it doesn't impact users… your way :)

I think it's already doable by installing two FreshRSS sharing the same DB, but it would definitely look "hacky". In the mean time, ease of installation is more important than scaling and solution 1 (more memory and CPU) is an easy way that I will go first. To be complete, there is another problem that a dedicated reader service could solve: IP-based blocking. It happens occasionally but for now I've been able to contact admin of these servers to allow my own server to fetch their feeds. So indeed, extracting a reader service would have benefits (to be balanced with mentioned drawbacks).

For the other services though, I'm afraid the idea would add complexity for almost zero impact.

**Comment by Alkarex** [comment]
@marienfressinaud Please give me a bit more info about the performance issue you observed (in another issue or by mail), it should be possible to fix it, assuming this was during a cron job.

@oupala Thanks for sharing. As Marien wrote, FreshRSS has a niche that is easy to deploy & maintain, light, widely known technologies, appropriate for self hosting (though scaling fine). Yes, other approaches / architectures / languages (I love C#) / frameworks would be better when prioritising other agendas. I personally have no problem adding new features to FreshRSS (only a lack of time, especially as I also want to help users and our community - a part where more people would be especially nice), and we do receive quite many pull requests, which seems to indicate that the contribution barrier is relatively low (which is good). But yes, some adjustments could be made, for instance if we want a JavaScript front-end at some point (which I think could be done with relatively few changes). At the moment, big architecture changes are not my priority though, as there is still quite a number of features I would like to implement (also for my own use 🙂).

Closing for now, but this does not mean that nothing will happen on the architecture side 🤓
